bot/exts/evergreen/music/app/api.py

```diff
@@ -0,0 +1,180 @@
+"""Module to handle API data for Music cog."""
+
+import asyncio
+import logging
+import time
+from dataclasses import dataclass, field
+from typing import Any, Dict, Optional, Tuple
+from urllib.parse import urlencode
+
+from aiohttp import ClientConnectionError, ClientResponseError, ClientSession
+
+from bot.constants import Tokens
+
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+class InvalidArgument(ValueError):
+    """Exception to raise when an argument is invalid."""
+
+    pass
+
+
+@dataclass
+class RequestInfo:
+    """Mutable data object to hold request info for methods."""
+
+    _time: Optional[float] = None
+
+    @property
+    def time(self) -> Optional[float]:
+        """Timestamp of current request."""
+        return self._time
+
+    @property
+    def time_since(self) -> float:
+        """
+        Time in seconds since last tracked request.
+
+        Time may not be exact, depending on when request is tracked.  Should not be
+          an issue, since values are used with caching and precision is not sensitive.
+        """
+        if self.time is None:
+            raise TypeError("No initial value for `time`. Call `update` before accessing other attributes.")
+        return time.time() - self.time
+
+    def update(self) -> None:
+        """
+        Update request info.
+
+        Call method via an API method every time a request is made.
+        """
+        self._time: float = time.time()
+
+
+@dataclass(frozen=True)
+class ApiMethod:
+    """
+    API method data.
+
+    An API method is immutable (frozen) by design, to avoid editing constant values.  However,
+      the time of when a request was made is used with caching.  A mutable protected
+      attribute can update request info while the other attributes remain immutable.
+    """
```

Comment:
Why is there a hanging indent?

Reply:
I believe it was there to indicate the text is continued from previous line, but was decided on Discord to remove so have done so :+1: